### PR TITLE
Webex integration

### LINKED ISSUE #2744: WebEx App
- Getting Started: https://developer.webex.com/docs/getting-started
- Webex integrations: https://developer.webex.com/docs/integrations

### PR DESCRIPTION
**Pull request by aar2dee2**
## What does this PR do?

Fixes #2744 
/claim #2744

**Environment**: Staging(main branch) / Production

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## How should this be tested?

- [ ] Test A
- [ ] Test B

## Checklist

- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by aar2dee2** [comment]
Hi,

I'm using the Zoom integration as a reference for building the Webex integration.
The Oauth is working:

https://user-images.githubusercontent.com/85004512/224246221-eb39ea34-d0dc-4904-884d-de16d0a0bf6f.mov

1/ I'm trying to test creating a meeting, but there's a Prisma error in the event types table. 
Screenshot 2023-03-09 at 6 22 05 PM

How should I fix this?

2/ I'm working on functionality for

- listing meetings
- creating a meeting
- updating a meeting
- deleting a meeting

Are any other functions required from the integration?


3/ Other TODOs

    - Screenshots
    - Tests
    -What else should be there?

**Comment by sean-brydon** [comment]
> Hi,
> 
> I'm using the Zoom integration as a reference for building the Webex integration. The Oauth is working:
> 
> 1/ I'm trying to test creating a meeting, but there's a Prisma error in the event types table. Screenshot 2023-03-09 at 6 22 05 PM
> 
> How should I fix this?
> 
> 2/ I'm working on functionality for
> 
> * listing meetings
> * creating a meeting
> * updating a meeting
> * deleting a meeting
> 
> Are any other functions required from the integration?
> 
> 3/ Other TODOs
> 
> ```
> - Screenshots
> - Tests
> -What else should be there?
> ```

Can you try running yarn clean &&  yarn dx in project root - looks like you might be missing a migration

**Comment by aar2dee2** [comment]
I'm running into an error while calculating the `utcOffset` from `event.startTime` and `event.organizer.timeZone`.
Here's the code for the `translateEvent` function inside the `VideoApiAdapter` file:
```ts
const translateEvent = (event: CalendarEvent) => {
    //To convert the Cal's CalendarEvent type to a webex meeting type
    console.log("event received in translateEvent", event);
    /** @link https://developer.webex.com/docs/api/v1/meetings/create-a-meeting */
   
    const utcOffset = dayjs(event.startTime, event.organizer.timeZone).utcOffset() / 60;
    console.log("calculated utcOffset", utcOffset);
    console.log("calculated start", dayjs(event.startTime).utcOffset(utcOffset).format());
    return {
      title: event.title,
      start: dayjs(event.startTime).utcOffset(utcOffset).format(),
      end: event.endTime,
      recurrence: event.recurrence, //Follows RFC 2445 https://www.ietf.org/rfc/rfc2445.txt, TODO check if needs conversion
      timezone: event.organizer.timeZone,
      agenda: event.description,
      enableJoinBeforeHost: true, //this is true in zoom's api, do we need it here?
      invitees: event.attendees.map((attendee) => ({
        email: attendee.email,
      })),
      sendEmail: true,
    };
  };
```

Here's the corresponding console log:
Screenshot 2023-03-16 at 7 10 40 PM

**Comment by aar2dee2** [comment]
After merging recent changes from calcom:main, I'm getting a Prisma error:
Screenshot 2023-03-20 at 11 40 25 AM

I've run `yarn clean`, `yarn` and `yarn workspace @calcom/prisma db-deploy`. 
The database is accessible in the Prisma studio.

The variables in `/packages/prisma/.env` are:
Screenshot 2023-03-20 at 12 53 35 PM

Also, do you use a specific app for taking screenshots (for using in the Calcom app store)?

**Comment by aar2dee2** [comment]
> After merging recent changes from calcom:main, I'm getting a Prisma error: Screenshot 2023-03-20 at 11 40 25 AM
> 
> I've run `yarn clean`, `yarn` and `yarn workspace @calcom/prisma db-deploy`. The database is accessible in the Prisma studio.
> 
> The variables in `/packages/prisma/.env` are: Screenshot 2023-03-20 at 12 53 35 PM
> 
> Also, do you use a specific app for taking screenshots (for using in the Calcom app store)?

nvm, fixed this. Needed to set `PRISMA_GENERATE_DATAPROXY=` as blank instead of `false`

**Comment by aar2dee2** [comment]
Creating a meeting:

If I create a meeting with `start` and `end` times in `UTC` and pass along the organizer's `timezone`, Webex throws a `404`.
(Error inducing) Request body:
```json
{
    "title":"30 Min Meeting between aar2dee2 and Bo Katan",
    "start":"2023-03-27T05:30:00Z",
    "end":"2023-03-27T06:00:00Z",
    "agenda":"",
    "timeZone": "Asia/Calcutta",
    "enableJoinBeforeHost":true,
    "invitees": [
              {"email":"thisistheway@gmail.com"}
          ],
    "sendEmail":true
}
```

Response body for above request:
```json
{
  "message": "timezone 'Asia/Calcutta' and timezone offset in '2023-03-21T05:00:00Z' do not match",
  "trackingId": "WEBEX-DEV-PORTAL_a000625b-de6a-4d9b-b40c-3d52ee522a05_2"
}
```

If I comment out the `timezone` parameter in `translateEvent` as below, the request is processed and the meeting is created.
```ts
const translateEvent = (event: CalendarEvent) => {
    //To convert the Cal's CalendarEvent type to a webex meeting type
    console.log("event received in translateEvent", event);
    /** @link https://developer.webex.com/docs/api/v1/meetings/create-a-meeting */
    //Required params - title, start, end
    //There are a ton of other options, what do we want to support?

    console.log("calculated start", dayjs(event.startTime).utc().format());
    return {
      title: event.title,
      start: dayjs(event.startTime).utc().format(),
      end: dayjs(event.endTime).utc().format(),
      recurrence: event.recurrence, //Follows RFC 2445 https://www.ietf.org/rfc/rfc2445.txt, TODO check if needs conversion
      // timezone: event.organizer.timeZone, // Comment this out for now
      agenda: event.description,
      enableJoinBeforeHost: true, //this is true in zoom's api, do we need it here?
      invitees: event.attendees.map((attendee) => ({
        email: attendee.email,
      })),
      sendEmail: true,
    };
  };
```
I assume it requires conversion to a `utcOffset` format. But if the api accepts `UTC` which Cal uses by default, can we avoid passing the timezone parameter altogether (it is not a required param for the Webex api)?

Screenshots from the successful request for creating a meeting, where timezone was not included in request.
Screenshot 2023-03-20 at 3 56 24 PM

Screenshot 2023-03-20 at 3 55 58 PM

**Comment by emrysal** [comment]
Hi @aar2dee2 - We have a similar approach in the office 365 translate event:
image

Would this work for your usecase as well? Regardless, you can assume startTime and endTime are always given in UTC, so if webex handles this for you just pass those, no utcOffset translation will be necessary

**Comment by aar2dee2** [comment]
> Hi @aar2dee2 - We have a similar approach in the office 365 translate event: image
> 
> Would this work for your usecase as well? Regardless, you can assume startTime and endTime are always given in UTC, so if webex handles this for you just pass those, no utcOffset translation will be necessary

Thanks for this! Proceeding without the `utcOffset` translation.


I have create and update meeting working correctly. For delete meeting, the `deleteMeeting` function in the Webex `VideoApiAdapter` is not called, resulting in the meeting getting canceled on Cal, but not on Webex.
https://www.loom.com/share/dcfad65245e34343922da192c734a6cb

I think this is because the `BookingReference` for the booking has the wrong `credentialId` for `webex_video`. (In the loom link below, the correct `credentialId` for `webex_video` is `13`, not `8`.
https://www.loom.com/share/97853e302cbf4f9581ab314ca7c43ba0

Is it possible this is happening because of how the app was named? I have the app name as `Webex` and slug as `webex` and had to do a fix in `/packages/app-store/index`. 
Screenshot 2023-03-22 at 10 08 00 PM

Also, do you have an integration todos list I can check against?

**Comment by aar2dee2** [comment]
Hello Cal team,

I was wondering if you've had a chance to review the changes I made recently. Please share any feedback or suggestions when it's convenient for you. Thank you!

**Comment by hariombalhara** [comment]
@aar2dee2 I will start reviewing it now. Thanks for the patience 🙏

File: README.md
Comment (hariombalhara) [inline README.md:454]: I am wondering should we move these README changes to the respective app?
We can still reference the app README here. That would keep the README small and not intimidate users who are not worried about various apps.

cc @zomars

File: packages/app-store/webex/DESCRIPTION.md
Comment (hariombalhara) [inline packages/app-store/webex/DESCRIPTION.md:3]: @aar2dee2 can you add proper images and logo for the app.

File: packages/types/environment.d.ts
Comment (hariombalhara) [inline packages/types/environment.d.ts:34]: We need to find a way to stop polluting the global env directly. I think every app should expose it's environment and if needed we can import that env here.

File: packages/app-store/index.ts
Comment (hariombalhara) [inline packages/app-store/index.ts:25]: I don't think we need to change that. So, we can remove this TODO.

File: packages/app-store/webex/pages/setup/index.tsx
Comment (hariombalhara) [inline packages/app-store/webex/pages/setup/index.tsx:1]: I don't think we need this file

**Review by hariombalhara — DISMISSED** [review]
Left some comments.

File: packages/app-store/webex/DESCRIPTION.md
Comment (aar2dee2) [reply packages/app-store/webex/DESCRIPTION.md:3]: Sure. Do you use a specific app to get the screenshots?

Comment (hariombalhara) [reply packages/app-store/webex/DESCRIPTION.md:3]: @aar2dee2 I haven't added screenshots by myself yet.  Simple screen grabs should be enough depicting  what the app can achieve,

File: apps/web/playwright/integrations.e2e.ts
Comment (aar2dee2) [reply apps/web/playwright/integrations.e2e.ts:350]: Do you want me to make any changes here?

File: packages/types/environment.d.ts
Comment (aar2dee2) [reply packages/types/environment.d.ts:34]: not making any changes here

File: packages/app-store/webex/DESCRIPTION.md
Comment (aar2dee2) [reply packages/app-store/webex/DESCRIPTION.md:3]: added

File: packages/app-store/index.ts
Comment (aar2dee2) [reply packages/app-store/index.ts]: done

File: apps/web/playwright/integrations.e2e.ts
Comment (hariombalhara) [reply apps/web/playwright/integrations.e2e.ts:350]: We can just remove these tests. Dead code.

File: packages/types/environment.d.ts
Comment (hariombalhara) [reply packages/types/environment.d.ts:34]: Yeah that was to create an issue out of it for future work.

**Comment by aar2dee2** [comment]
Hi @hariombalhara 
I've added screenshots & icon. The issue with permissions should be fixed now. We only need `meetings:schedules_read` and `meeting:schedules_write` permissions. Have updated the `readme` for this.
I was able to create, update & delete meetings. The type error should be fixed.

**Comment by hariombalhara** [comment]
I am still getting the error related to scope.
image

I have set the scopes correctly.
Screenshot 2023-05-01 at 7 49 50 PM

Have tried to uninstall and reinstall the app. I even created a new Integration App in Webex by following the steps in README.

@aar2dee2  You can book me [here](https://cal.com/hariom/30min) if you want and we can together solve this on call.

**Review by hariombalhara — DISMISSED** [review]
Requesting changes as I can't get the app to work.

**Comment by aar2dee2** [comment]
> Requesting changes as I can't get the app to work.

Hi @hariombalhara 
The error is likely because I missed adding the `auth` callback uri in the `README` file. Updated this now.

I created a new integration and it works. See [video](https://www.loom.com/share/f370680c06094f2d89f3306631dc8612)
Have booked time on your calendar as well, incase needed.

**Review by hariombalhara — APPROVED** [review]
LGTM now. 

Tests:
- Verified that the meeting is successfully created in webex after following the steps.
- Meeting time in Webex is correct

File: packages/features/bookings/lib/handleCancelBooking.ts
Comment (hariombalhara) [reply packages/features/bookings/lib/handleCancelBooking.ts:242]: Yet to verify this flow. So, blocking merge.

**Review by hariombalhara — CHANGES_REQUESTED** [review]
Yet to verify a change in the core booking flow

File: packages/prisma/enums/index.ts
Comment (hariombalhara) [inline packages/prisma/enums/index.ts:1]: This shouldn't be committed. It is generated using `yarn prisma migrate dev` command. Or if you use `yarn dx` that command automatically runs.
Screenshot 2023-05-10 at 6 22 08 PM

File: yarn.lock
Comment (aar2dee2) [reply yarn.lock:1]: done!

thanks for helping resolve this on call!

File: packages/features/bookings/lib/handleCancelBooking.ts
Comment (hariombalhara) [reply packages/features/bookings/lib/handleCancelBooking.ts:242]: @alannnc I don't think this change is required. But this logic is confusing, can you confirm if this change is correct.

Also @aar2dee2 I was able to successfully delete the corresponding webex meeting when last attendee cancelled.

So, for now can you remove this change. We can followup with  this fix later if needed,

File: packages/app-store/webex/lib/VideoApiAdapter.ts
Comment (hariombalhara) [inline packages/app-store/webex/lib/VideoApiAdapter.ts:314]: https://github.com/calcom/cal.com/blob/main/packages/app-store/larkcalendar/api/events.ts#L52
Let's use logger for all console.logs

File: packages/features/bookings/lib/handleCancelBooking.ts
Comment (hariombalhara) [inline packages/features/bookings/lib/handleCancelBooking.ts:507]: I reverted this entire file and was still able to cancel the webex meeting by cancelling the booking

File: packages/app-store/webex/lib/VideoApiAdapter.ts
Comment (hariombalhara) [inline packages/app-store/webex/lib/VideoApiAdapter.ts:24]: We just need `start` and `end`. So, we can remove all other props. That should reduce this file's size and simplify it.

File: packages/app-store/webex/lib/VideoApiAdapter.ts
Comment (hariombalhara) [inline packages/app-store/webex/lib/VideoApiAdapter.ts:148]: These are also not needed.

File: packages/features/bookings/lib/handleCancelBooking.ts
Comment (aar2dee2) [reply packages/features/bookings/lib/handleCancelBooking.ts:507]: @hariombalhara 

the cancel booking behavior is not consistent for me. I reverted the file to the original code and am running into errors again while canceling the booking. The `videoAdapter` is not picked correctly from credentials. This results in the meeting getting deleted on cal, but not on Webex. I added logs in a couple of places to show this.

Screenshot 2023-05-11 at 9 36 28 AM
Screenshot 2023-05-11 at 9 36 16 AM

Comment (hariombalhara) [reply packages/features/bookings/lib/handleCancelBooking.ts]: Thanks @aar2dee2. Looks like some weird problem. But as I see it, it has nothing to do with your app. Let's keep it reverted for this PR.
I would let @alannnc or @joeauyeung take care of it as they have more knowledge about this flow and this should be a problem with all video apps including Zoom in this case.

Comment (aar2dee2) [reply packages/features/bookings/lib/handleCancelBooking.ts]: okay. thanks!

any other recommended changes?

**Review by hariombalhara — APPROVED** [review]
LGTM now !! It is in a mergeable state. I see that console.logs are still there instead of logger but that can be a followup PR.

Regarding that change in handleCancelBooking, that's a core change which shoould be tackled separately.

**Comment by aar2dee2** [comment]
> 

thank you!!

**Pull request merged**
